### DISCUSSION TITLE
Service names not shown in HomeKit Controller

### DISCUSSION THREAD

**Discussion by flohse123** [HomeSpan/HomeSpan#485] [Q&A] [answered]
Hi, I am just getting started working with your library and I am loving the work you did especially in documenting the different possibilities.

However in Example 11 my iPhone is not showing the names of services as described in the commentaries: The fourth Accessory (Ceiling Lights) is working as expected. Strangly in the fifth and sixth Accessory the names of the LightBulbs are shown as "Fan with Lights" or "Multi-Function Fan" respectively.

**Comment by HomeSpan** [comment, marked as answer]
Thanks for raising this.  It must be a ramification of the latest iOS update (16.2).  Apple seems to change the naming conventions every few iterations.  I'll experiment and see if I can determine the new rules of the road.

**Comment by chrwh** [comment]
Hi,
Has this issue been solved? I'm not getting names either :-(
Best regards,
Christian

**Comment by HomeSpan** [comment]
It seems that Apple may not allow the same level of naming granularity that it formally did in the Home App.  However, as long as you provide a name for each separate Accessory in your device, this name should show up, except for the first Accessory, which will have a default name set equal to your device name.  You can of course change these directly in the Home App at any point.

**Comment by chrwh** [comment]
Hmm ok, it's just that I have 14 switches in one Accessory and it's a nightmare to name the all via Home App

**Comment by HomeSpan** [comment]
@chrwh, I found a solution for this!

A while ago I added a new Characteristic called *ConfiguredName* to HomeSpan.  Apple apparently designed it to work with *InputSource* Services that are linked to a *Television* Service so you can create custom names for TV inputs, like "DVD", "Cable Box," etc.

However, *ConfiguredName* also seemed to be much more useful than *Name* since it was writeable as well as readable.  This meant that any changes to the name of an *InputSource* in the Home App would trigger an update in HomeSpan, just like changes do for any other Characteristic.

At the time I tried using *ConfiguredName* instead of *Name* for other Services, but HomeKit ignored it when used with other Services.  But your question prompted me to do some more web research, and others reported that starting with the release of iOS 17, *ConfiguredName* now works with other Services (when I did my previous testing Apple was still on iOS 16).  It works great and does exactly what one would expect:  whatever you name your Service in HomeSpan using *ConfiguredName* is transferred to the Home App when pairing, which means you **do not** need to rename them manually in the Home App.  I tested this in HomeSpan with the following sketch that defines two separate dual-light/dual-switch Accessories:

```C++
#include "HomeSpan.h"

void setup() {
 
  Serial.begin(115200);

  homeSpan.setLogLevel(2);

  homeSpan.begin(Category::Lighting,"HomeSpan Test");
  
   new SpanAccessory();                       // start with a Bridge Accessory
    new Service::AccessoryInformation();  
      new Characteristic::Identify(); 

// First dual-light/dual-switch Accessory

   new SpanAccessory();
    new Service::AccessoryInformation();  
      new Characteristic::Identify();  
      new Characteristic::Name("Access-1");      // the AccessoryInformation Service seems to still require Name(), and will ignore ConfiguredName()

    new Service::LightBulb();
      new Characteristic::On(0);
      new Characteristic::ConfiguredName("Main Light");     // here we use ConfiguredName() instead of Name().  This is properly reflected in Home App during pairing

    new Service::LightBulb();
      new Characteristic::On(0);
      new Characteristic::ConfiguredName("Night Light");   
      
    new Service::Switch();
      new Characteristic::On(0);
      new Characteristic::ConfiguredName("Switch-A");

    new Service::Switch();
      new Characteristic::On(0);
      new Characteristic::ConfiguredName("Switch-B");

// Second dual-light/dual-switch Accessory

   new SpanAccessory();
    new Service::AccessoryInformation();  
      new Characteristic::Identify();  
      new Characteristic::Name("Access-2");               // note as above we use Name() instead of ConfiguredName() for the AccessoryInformation Service

    new Service::LightBulb();
      new Characteristic::On(0);
      new Characteristic::ConfiguredName("Aux Light");  // but then we once again use ConfiguredName() for all the other Services

    new Service::LightBulb();
      new Characteristic::On(0);
      new Characteristic::ConfiguredName("Night Light");
      
    new Service::Switch();
      new Characteristic::On(0);
      new Characteristic::ConfiguredName("Switch-A");

    new Service::Switch();
      new Characteristic::On(0);
      new Characteristic::ConfiguredName("Switch-B");
}

//////////////////////////////////////

void loop(){
  homeSpan.poll();
}
```

When I paired HomeSpan with the Home App using the above sketch, two new Accessories were added, with names "Access-1" and "Access-2."  More so, the names of the LightBulbs and Switches in each Accessory properly reflected the names in the sketch, exactly as one would hope.

A few caveats to note:

* I have not yet tested *ConfiguredName* with any other Services except for the LightBulb and Switch.
* Though the Home App seems to properly interpret and use the *ConfiguredName* during pairing, whether or not it actually DISPLAYS this name under the slider for a LightBulb or Switch depends on if you have multiple Services of the same type.  For example, if you eliminate either one of the LightBulb Services, the Home App will NOT display any name below the remaining LightBulb Service, though it displays the names of both the Switches.  If you navigate to the Settings page in the Home App you can see that the LightBulb name is indeed present -- it's just not displayed.  I guess Apple figures they only need to display the names of the Service when there is ambiguity because an Accessory contains multiple Services of the same type.  Since the LightBulb "slider" is different from the Switch "slider" there is no confusion as to which is the LightBulb slider.  I wish Apple stopped building this type of logic into the Home App and relied on the judgement of those creating the devices to determine if and when to display names.
* The very first Accessory still seems to ALWAYS inherit the name of the Device (in this example, the device name is "HomeSpan Test").  By using a Bridge as the first Accessory, this allows the other Accessory names to appear correctly.  If you eliminate the first Bridge Accessory in the above example, HomeSpan will still show two Accessory tiles, but the first will be named "HomeSpan Test" and the second will be named "Access-2".

Please give this a try and let me know how it goes.  If it works I'll update the documentation and Example 7 to use *ConfiguredName* instead of *Name*.

**Comment by chrwh** [comment]
Hi,
Thanks for the great work. I have tested it and it seems to be working as intended. I have also tested it in my own sketch and it also seems to be working as it should :-)
Thanks and best regards,
Christian
